### ISSUE TITLE
CASSGO-23 Make Session immutable

### ISSUE THREAD

**Issue by Zariel** [apache/cassandra-gocql-driver#357]
Currently Session has some fields which can be changed, such as Cons, Tracing, and some others. This requires there to be a `RWMutex` which needs to be acquired when executing a query to copy the options.

I propose we make Session immutable and remove this lock.

**Comment by 0x6e6562** [comment]
It's always nice when a lock can be feasibly removed :-)

**Comment by Zariel** [comment]
It would mean some things get moved to the config and become immutable session wide, such as the tracer.

**Comment by tux21b** [comment]
I think it would be nice to have a separation between a `Session` and the `Cluster`. It might make sense to enable tracing (or the default consistency, etc.) only for one part of the application. The initial idea (I am not yet sure how far the current gocql API has diverged) was to have multiple sessions (with different session specific configs) that can use the same underlying cluster. So that each component of a larger application can configure the session differently (for logging / tracing purposes for example). I'm not sure about the popularity of this feature though.

But I agree that it would also be nice to have an immutable Session object. We could archive both by providing methods like `func (s *Session) WithTracing(t Tracer) *Session` that return a new copy of the Session.

**Comment by Zariel** [comment]
Interesting, I've never come across that use case, in general I always assumed that an application would have a single session. Given that you can change tracing on a per query basis Im not sure what use cases there are to do it session wide (maybe set tracing and write the tracing output to a `http.ResposneWriter`?).

Given that #375 essentially made NewCluster a wrapper for `&ClusterConfig{Hosts: }` Im not sure what the current separation between Cluster and Session is used for, as now `NewSession` does all the setup.

**Comment by 0x6e6562** [comment]
I wonder how many apps in practice actually use more than one session instance (other than to control tracing in a more differentiated fashion)?

**Comment by joao-r-reis** [comment]
I'm considering adding this to the scope of 2.0.0 but I'm a bit concerned about the impact of the breaking changes. Maybe we can just make these functions create a copy of `Session` like @tux21b mentioned

**Comment by jameshartig** [comment]
I'm concerned about making them return a copy because the only time you call those methods is after the session is created. If it returns a copy then do you need to close the old session? it also means unnecessary connections if you only want 1 session.

Can we instead change `NewSession` to accept a `SessionConfig` that has a `ClusterConfig` inside of it? Then we can add fields to `SessionConfig` for whatever we need to set and remove the setter methods.

**Comment by joao-r-reis** [comment]
> I'm concerned about making them return a copy because the only time you call those methods is after the session is created. If it returns a copy then do you need to close the old session? it also means unnecessary connections if you only want 1 session.

I was thinking of shallow copying so you would still keep the same connections with the new session but it might become confusing for users to decide which session object they should close and when... 

> Can we instead change NewSession to accept a SessionConfig that has a ClusterConfig inside of it? Then we can add fields to SessionConfig for whatever we need to set and remove the setter methods.

You would still have to use a mutex to manage reading and writing these fields right? The main goal of this issue was to get rid of this mutex.

**Comment by jameshartig** [comment]
> You would still have to use a mutex to manage reading and writing these fields right? The main goal of this issue was to get rid of this mutex.

`NewSession(ClusterConfig)` right now will copy it so there's no mutex needed since the caller can't mutate the copied config. Same thing if it was `NewSession(ClusterConfig, SessionConfig)`. The reason why I think we'd need to accept both is because we'd want to make `(*ClusterConfig).CreateSession(SessionConfig)`. It's also a bit weird there's 2 different ways to make a session and they're almost identical.

**Comment by joao-r-reis** [comment]
I don't understand what you're suggesting as an alternative to returning a copy of the session, can you provide a small example?

**Comment by ribaraka** [comment]
> I'm considering adding this to the scope of 2.0.0 but I'm a bit concerned about the impact of the breaking changes. Maybe we can just make these functions create a copy of `Session` like [@tux21b](https://github.com/tux21b) mentioned

@jameshartig is right. Creating copies of the `Session` can lead to resource management challenges. Each `Session` maintains its own connection pool to a cluster. Therefore, duplicating sessions can result in redundant connections, increasing the load on both the client and the cluster. Moreover, managing the lifecycle of these duplicated sessions becomes complex, raising the risk of resource leaks if sessions are not properly closed. 
Why not leverage per-query configuration methods, such as Query.Consistency and Query.Trace? They offer flexibility without requiring the creation of multiple sessions, aren't they? 

I suggest continue using a single `Session` instance for the application's lifecycle, utilizing per-query configurations to handle specific requirements. Remove setters and keep session scoped settings to `ClusterConfig`.

I actually like the solution by @Zariel made in https://github.com/apache/cassandra-gocql-driver/pull/1290 it just need to be reviewed and updated.

**Comment by joao-r-reis** [comment]
Yeah I'm ok with removing these methods as long as there is a way to configure these properties via query / batch properties and via clusterconfig fields.

I think @Zariel is not active in the project so we can just go ahead and create a PR from scratch to remove these methods I think

**Comment by ribaraka** [comment]
OK! I'll do it

**Comment by joao-r-reis** [comment]
v2.0.0-rc1 with the fix for this has been released, it would be great if users could try it out and provide feedback on it before 2.0.0 GA
